File: frontend/scripts/import-colosseum-builders.py
#!/usr/bin/env python3
"""
Import colosseum_builders.csv into the builders table.

Usage:
    python3 scripts/import-colosseum-builders.py

Generates:
    migrations/import-colosseum-builders.sql (single file, for local)
    migrations/import-colosseum-builders-NNN.sql (batches, for remote)

Apply:
    npx wrangler d1 execute <your-d1-database-name> --local --file=migrations/import-colosseum-builders.sql
"""

import csv
import json
import uuid
import os
from datetime import datetime, timezone

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
CSV_PATH = os.path.join(SCRIPT_DIR, "..", "colosseum_builders.csv")
OUT_DIR = os.path.join(SCRIPT_DIR, "..", "migrations")
OUT_PATH = os.path.join(OUT_DIR, "import-colosseum-builders.sql")

BATCH_SIZE = 500
NOW = datetime.now(timezone.utc).isoformat()


def split_list(val: str) -> list:
    if not val or not val.strip():
        return []
    return [s.strip() for s in val.split(",") if s.strip()]


def escape_sql(s: str) -> str:
    return s.replace("'", "''")


def main():
    with open(CSV_PATH, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)

        statements = []
        skipped = 0

        for row in reader:
            username = (row.get("username") or "").strip()
            if not username:
                skipped += 1
                continue

            builder_data = {
                "username": username,
                "display_name": (row.get("display_name") or username).strip(),
                "avatar_url": "",
                "position": (row.get("position") or row.get("current_position_api") or "").strip(),
                "city": (row.get("city_api") or row.get("location") or "").strip(),
                "about": (row.get("about") or "").strip(),
                "skills": split_list(row.get("skills", "")),
                "i_am_a": split_list(row.get("i_am_a") or row.get("roles_api") or ""),
                "looking_for": split_list(row.get("looking_for", "")),
                "interested_in": split_list(row.get("interested_in", "")),
                "languages": split_list(row.get("languages") or row.get("languages_api") or ""),
                "looking_for_teammates_text": (row.get("looking_for_teammates_text") or "").strip(),
                "is_student": (row.get("is_student") or "").strip().lower() == "true",
                "twitter_url": (row.get("twitter") or "").strip(),
                "github_url": (row.get("github") or "").strip(),
                "telegram_url": (row.get("telegram") or "").strip(),
                "wallet_address": "",
                "claimed": False,
                "source": "colosseum",
                "created_at": NOW,
            }

            bid = str(uuid.uuid4())
            json_str = escape_sql(json.dumps(builder_data, ensure_ascii=False))
            statements.append(
                f"INSERT OR IGNORE INTO builders (id, data) VALUES ('{bid}', '{json_str}');"
            )

    imported = len(statements)
    print(f"Parsed {imported} builders, {skipped} skipped (no username)")

    # Verify first entry
    if statements:
        first_json = statements[0].split("', '")[1].rstrip("');").replace("''", "'")
        d = json.loads(first_json)
        print(f"\nFirst builder check:")
        print(f"  username: {d['username']}")
        print(f"  city: {d['city']}")
        print(f"  about: {d['about'][:60]}...")
        print(f"  skills: {d['skills'][:3]}")
        print(f"  i_am_a: {d['i_am_a'][:3]}")
        print(f"  twitter: {d['twitter_url']}")

    cleanup = "DELETE FROM builders WHERE json_extract(data, '$.source') = 'colosseum';"

    # Single file for local
    with open(OUT_PATH, "w", encoding="utf-8") as f:
        f.write(f"-- Auto-generated: import colosseum builders\n")
        f.write(f"-- {imported} builders, generated at {NOW}\n\n")
        f.write(f"-- Purge old colosseum imports first\n")
        f.write(f"{cleanup}\n\n")
        f.write("\n".join(statements))
        f.write("\n")

    # Batch files for remote
    total_batches = (imported + BATCH_SIZE - 1) // BATCH_SIZE
    for b in range(total_batches):
        batch = statements[b * BATCH_SIZE : (b + 1) * BATCH_SIZE]
        batch_num = f"{b + 1:03d}"
        batch_path = os.path.join(OUT_DIR, f"import-colosseum-builders-{batch_num}.sql")
        with open(batch_path, "w", encoding="utf-8") as f:
            f.write(f"-- Colosseum builders batch {b + 1}/{total_batches}\n")
            f.write(f"-- {len(batch)} builders\n\n")
            if b == 0:
                f.write(f"-- Purge old colosseum imports first\n")
                f.write(f"{cleanup}\n\n")
            f.write("\n".join(batch))
            f.write("\n")

    print(f"\nDone! {imported} builders, {total_batches} batch files")
    print(f"\nApply locally:")
    print(f"  npx wrangler d1 execute <your-d1-database-name> --local --file=migrations/import-colosseum-builders.sql")
    print(f"\nApply remote (batches):")
    print(f'  for f in migrations/import-colosseum-builders-*.sql; do npx wrangler d1 execute <your-d1-database-name> --file="$f"; done')


if __name__ == "__main__":
    main()

File: frontend/scripts/import-colosseum-builders.ts
/**
 * Import colosseum_builders.csv into the builders table.
 *
 * Usage:
 *   npx tsx scripts/import-colosseum-builders.ts
 *
 * This generates migrations/import-colosseum-builders.sql
 * Then apply with:
 *   npx wrangler d1 execute <your-d1-database-name> --local --file=migrations/import-colosseum-builders.sql
 *   npx wrangler d1 execute <your-d1-database-name> --file=migrations/import-colosseum-builders.sql
 */

import { readFileSync, writeFileSync } from "fs"
import { randomUUID } from "crypto"
import { resolve, dirname } from "path"
import { fileURLToPath } from "url"

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const CSV_PATH = resolve(__dirname, "../colosseum_builders.csv")
const OUT_PATH = resolve(__dirname, "../migrations/import-colosseum-builders.sql")

// Simple CSV parser that handles quoted fields with commas inside
function parseCSVLine(line: string): string[] {
  const fields: string[] = []
  let current = ""
  let inQuotes = false

  for (let i = 0; i < line.length; i++) {
    const ch = line[i]
    if (ch === '"') {
      if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
        current += '"'
        i++ // skip escaped quote
      } else {
        inQuotes = !inQuotes
      }
    } else if (ch === "," && !inQuotes) {
      fields.push(current)
      current = ""
    } else {
      current += ch
    }
  }
  fields.push(current)
  return fields
}

function splitList(val: string): string[] {
  if (!val || !val.trim()) return []
  return val
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean)
}

function escapeSQL(s: string): string {
  return s.replace(/'/g, "''")
}

// Split CSV content into logical rows, handling multiline quoted fields
function splitCSVRows(content: string): string[] {
  const rows: string[] = []
  let current = ""
  let inQuotes = false

  for (let i = 0; i < content.length; i++) {
    const ch = content[i]
    if (ch === '"') {
      if (inQuotes && i + 1 < content.length && content[i + 1] === '"') {
        current += '""'
        i++
      } else {
        inQuotes = !inQuotes
      }
      current += ch
    } else if (ch === "\n" && !inQuotes) {
      rows.push(current)
      current = ""
    } else {
      current += ch
    }
  }
  if (current.trim()) rows.push(current)
  return rows
}

function main() {
  const raw = readFileSync(CSV_PATH, "utf-8")
  const rows = splitCSVRows(raw)
  const header = rows[0]
  const cols = parseCSVLine(header)

  console.log(`Columns: ${cols.join(", ")}`)
  console.log(`Total logical rows (including header): ${rows.length}`)

  const colIndex: Record<string, number> = {}
  cols.forEach((c, i) => {
    colIndex[c.trim()] = i
  })

  const now = new Date().toISOString()
  const statements: string[] = []
  let imported = 0
  let skipped = 0

  for (let i = 1; i < rows.length; i++) {
    const line = rows[i].trim()
    if (!line) continue

    const fields = parseCSVLine(line)
    if (fields.length < cols.length) {
      skipped++
      continue
    }
    const get = (col: string) => (fields[colIndex[col]] || "").trim()

    const username = get("username")
    if (!username) {
      skipped++
      continue
    }

    const builderData = {
      username,
      display_name: get("display_name") || username,
      avatar_url: "",
      position: get("position") || get("current_position_api") || "",
      city: get("city_api") || get("location") || "",
      about: get("about") || "",
      skills: splitList(get("skills")),
      i_am_a: splitList(get("i_am_a") || get("roles_api")),
      looking_for: splitList(get("looking_for")),
      interested_in: splitList(get("interested_in")),
      languages: splitList(get("languages") || get("languages_api")),
      looking_for_teammates_text: get("looking_for_teammates_text") || "",
      is_student: get("is_student")?.toLowerCase() === "true",
      twitter_url: get("twitter") || "",
      github_url: get("github") || "",
      telegram_url: get("telegram") || "",
      wallet_address: "",
      claimed: false,
      source: "colosseum" as const,
      created_at: now,
    }

    const id = randomUUID()
    const json = escapeSQL(JSON.stringify(builderData))

    statements.push(
      `INSERT OR IGNORE INTO builders (id, data) VALUES ('${id}', '${json}');`
    )
    imported++
  }

  const CLEANUP = "DELETE FROM builders WHERE json_extract(data, '$.source') = 'colosseum';"

  // Split into batch files of BATCH_SIZE statements each (D1 has size limits)
  const BATCH_SIZE = 500
  const totalBatches = Math.ceil(statements.length / BATCH_SIZE)

  for (let b = 0; b < totalBatches; b++) {
    const batch = statements.slice(b * BATCH_SIZE, (b + 1) * BATCH_SIZE)
    const batchNum = String(b + 1).padStart(3, "0")
    const batchPath = resolve(__dirname, `../migrations/import-colosseum-builders-${batchNum}.sql`)
    const lines = [
      `-- Colosseum builders batch ${b + 1}/${totalBatches}`,
      `-- ${batch.length} builders`,
      "",
    ]
    // Only the first batch cleans up old data
    if (b === 0) {
      lines.push("-- Purge old colosseum imports first")
      lines.push(CLEANUP)
      lines.push("")
    }
    lines.push(...batch, "")
    writeFileSync(batchPath, lines.join("\n"), "utf-8")
  }

  // Also write a single combined file for local use
  const fullSql = [
    "-- Auto-generated: import colosseum builders",
    `-- ${imported} builders, generated at ${now}`,
    "",
    "-- Purge old colosseum imports first",
    CLEANUP,
    "",
    ...statements,
    "",
  ].join("\n")
  writeFileSync(OUT_PATH, fullSql, "utf-8")

  console.log(`\nDone! ${imported} builders exported, ${skipped} skipped (no username)`)
  console.log(`Split into ${totalBatches} batch files (${BATCH_SIZE} per batch)`)
  console.log(`\nApply locally (single file):`)
  console.log(`  npx wrangler d1 execute <your-d1-database-name> --local --file=migrations/import-colosseum-builders.sql`)
  console.log(`\nApply remote (batches):`)
  console.log(`  for f in migrations/import-colosseum-builders-*.sql; do npx wrangler d1 execute <your-d1-database-name> --file="$f"; done`)
}

main()
